src/gadget/sha256/table16/gates.rs

```diff
@@ -0,0 +1,116 @@
+use crate::arithmetic::FieldExt;
+use crate::plonk::Expression;
+
+pub struct Gate<F: FieldExt>(pub Expression<F>);
+
+impl<F: FieldExt> Gate<F> {
+    const ONES: Expression<F> = Expression::Ones();
+
+    // Helper gates
+    fn lagrange_interpolate(
+        var: Expression<F>,
+        points: Vec<u16>,
+        evals: Vec<u32>,
+    ) -> (F, Expression<F>) {
+        assert_eq!(points.len(), evals.len());
+        let deg = points.len();
+
+        fn factorial(n: u64) -> u64 {
+            if n < 2 {
+                1
+            } else {
+                n * factorial(n - 1)
+            }
+        }
+
+        // Scale the whole expression by factor to avoid divisions
+        let factor = factorial((deg - 1) as u64);
+
+        let numerator = |var: Expression<F>, eval: u32, idx: u64| {
+            let mut expr = Self::ONES;
+            for i in 0..deg {
+                let i = i as u64;
+                if i != idx {
+                    expr = expr * (Self::ONES * (-F::one()) * F::from_u64(i) + var.clone());
```

Comment:
Consider adding a `From` impl so that we can say
```suggestion
                    expr = expr * ((-F::one()) * F::from_u64(i) + var.clone()).into();
```